### PR TITLE
Feature: adding mongo to postgresql etl

### PR DESCRIPTION
**Pull request by eapg**
* Adding mongo to postgresql orderStatusHistory mapper
* Adding mongo to postgresql etl

### CODE REVIEW

File: src/main/java/com/ea/restaurant/etl/MongoToPostgresqlOrderStatusHistory.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import com.ea.restaurant.service.MongoOrderStatusHistoryService;
+import com.ea.restaurant.service.OrderStatusHistoryService;
+import com.ea.restaurant.util.OrderStatusHistoryMapper;
+import com.ea.restaurant.util.OrderStatusHistoryUtil;
+import java.util.ArrayList;
+import java.util.List;
+import org.bson.types.ObjectId;
+import org.springframework.scheduling.annotation.Scheduled;
+
+public class MongoToPostgresqlOrderStatusHistory {
```
Comment (reynoldmorel) [inline src/main/java/com/ea/restaurant/etl/MongoToPostgresqlOrderStatusHistory.java:15]: Could you create a class to abstract how an ETL works?

Example:
```
public interface Etl<S, T> {

  S extract();
  
 T extract(S s);

 void load(T t);

}
```

Every ETL should implement this interface.

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/main/java/com/ea/restaurant/etl/MongoToPostgresqlOrderStatusHistory.java (hunk 1 above)
Comment (eapg) [reply src/main/java/com/ea/restaurant/etl/MongoToPostgresqlOrderStatusHistory.java:15]: Done.

File: src/main/java/com/ea/restaurant/etl/MongoToPostgresqlOrderStatusHistory.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  private List<MongoOrderStatusHistory> extractData() {
+    return this.mongoOrderStatusHistoryService.findUnProcessedOrderStatusHistories();
+  }
+
+  private List<OrderStatusHistory> transformData(List<MongoOrderStatusHistory> extractedData) {
+
+    var transformedData = new ArrayList<OrderStatusHistory>();
+    for (MongoOrderStatusHistory mongoOrderStatusHistory : extractedData) {
```
[the remark this answers is not in the document]
Comment (eapg) [reply src/main/java/com/ea/restaurant/etl/MongoToPostgresqlOrderStatusHistory.java:41]: Done.

File: src/main/java/com/ea/restaurant/etl/MongoToPostgresqlOrderStatusHistory.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      transformedData.add(
+          OrderStatusHistoryMapper.mapMongoOrderStatusToPostgresqlOrderStatus(
+              mongoOrderStatusHistory));
+    }
+    return transformedData;
+  }
+
+  private void loadData(List<OrderStatusHistory> transformedData) {
+    var mongoUuids = new ArrayList<ObjectId>();
```
[the remark this answers is not in the document]
Comment (eapg) [reply src/main/java/com/ea/restaurant/etl/MongoToPostgresqlOrderStatusHistory.java:50]: Done.

File: src/main/java/com/ea/restaurant/util/OrderStatusHistoryUtil.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  public static List<OrderStatusHistory> updateLastOrderStatusHistory(
+      List<OrderStatusHistory> lastOrderStatusHistoriesInBd,
+      List<OrderStatusHistory> lastOrderStatusHistory,
+      Long updatedBy) {
+    var updatedOrderStatusHistories = new ArrayList<OrderStatusHistory>();
+
+    for (OrderStatusHistory orderStatusHistory : lastOrderStatusHistory) {
+
+      if (orderStatusHistory.getFromStatus() != OrderStatus.NEW_ORDER) {
+        var orderStatusHistoryToBeUpdated =
```
[the remark this answers is not in the document]
Comment (eapg) [reply src/main/java/com/ea/restaurant/util/OrderStatusHistoryUtil.java:18]: Done.

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/main/java/com/ea/restaurant/etl/impl/MongoToPostgresqlOrderStatusHistory.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import com.ea.restaurant.service.MongoOrderStatusHistoryService;
+import com.ea.restaurant.service.OrderStatusHistoryService;
+import com.ea.restaurant.util.OrderStatusHistoryMapper;
+import com.ea.restaurant.util.OrderStatusHistoryUtil;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import org.bson.types.ObjectId;
+
+public class MongoToPostgresqlOrderStatusHistory
```
[the remark this answers is not in the document]
Comment (eapg) [reply src/main/java/com/ea/restaurant/etl/impl/MongoToPostgresqlOrderStatusHistory.java:16]: Done.

File: src/main/java/com/ea/restaurant/etl/impl/MongoToPostgresqlOrderStatusHistory.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  private final MongoOrderStatusHistoryService mongoOrderStatusHistoryService;
+
+  public MongoToPostgresqlOrderStatusHistory(
+      OrderStatusHistoryService orderStatusHistoryService,
+      MongoOrderStatusHistoryService mongoOrderStatusHistoryService) {
+    this.orderStatusHistoryService = orderStatusHistoryService;
+    this.mongoOrderStatusHistoryService = mongoOrderStatusHistoryService;
+  }
+
+  public void process() {
```
[the remark this answers is not in the document]
Comment (eapg) [reply src/main/java/com/ea/restaurant/etl/impl/MongoToPostgresqlOrderStatusHistory.java:29]: Done.

File: src/main/java/com/ea/restaurant/etl/Etl.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package com.ea.restaurant.etl;
+
+import java.util.List;
+
+public interface Etl<E, S> {
```
[the remark this answers is not in the document]
Comment (eapg) [reply src/main/java/com/ea/restaurant/etl/Etl.java:5]: I use E for entity1 and S for entity2 in this case of an ETL the entity1 corresponds to extracted entity type, and the entity2 for the transformed entity type.

File: src/main/java/com/ea/restaurant/etl/Etl.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+  List<S> extract();
+
+  List<E> transform(List<S> list);
```
[the remark this answers is not in the document]
Comment (eapg) [reply src/main/java/com/ea/restaurant/etl/Etl.java:8]: Done.

File: src/main/java/com/ea/restaurant/etl/Etl.java [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+
+  void load(List<E> list);
```
[the remark this answers is not in the document]
Comment (eapg) [reply src/main/java/com/ea/restaurant/etl/Etl.java:10]: Done.

File: src/main/java/com/ea/restaurant/etl/Etl.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (reynoldmorel) [reply src/main/java/com/ea/restaurant/etl/Etl.java:5]: I don't think entity 2 is a valid name, Could you use maybe the concept of S = source and T = target?

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/main/java/com/ea/restaurant/etl/Etl.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (eapg) [reply src/main/java/com/ea/restaurant/etl/Etl.java:5]: Done.

File: src/main/java/com/ea/restaurant/repository/MongoOrderStatusHistoryRepository.java [hunk 10]
```diff
@@ -12,5 +13,5 @@
     extends MongoRepository<MongoOrderStatusHistory, String> {
   List<MongoOrderStatusHistory> findAllByEtlStatus(EtlStatus etlStatus);
 
-  List<MongoOrderStatusHistory> findAllByIdIn(List<ObjectId> ids);
+  List<MongoOrderStatusHistory> findAllByIdIn(HashSet<ObjectId> ids);
```
Comment (reynoldmorel) [inline src/main/java/com/ea/restaurant/repository/MongoOrderStatusHistoryRepository.java:16]: Use the `Set` interface instead

File: src/main/java/com/ea/restaurant/service/MongoOrderStatusHistoryService.java [hunk 11]
```diff
@@ -1,11 +1,12 @@
...
 import com.ea.restaurant.document.MongoOrderStatusHistory;
+import java.util.HashSet;
 import java.util.List;
 import org.bson.types.ObjectId;
 
 public interface MongoOrderStatusHistoryService {
   List<MongoOrderStatusHistory> findUnProcessedOrderStatusHistories();
 
-  void updateBatchToProcessed(List<ObjectId> ids);
+  void updateBatchToProcessed(HashSet<ObjectId> ids);
```
Comment (reynoldmorel) [inline src/main/java/com/ea/restaurant/service/MongoOrderStatusHistoryService.java:11]: Use the Set interface instead

File: src/main/java/com/ea/restaurant/service/impl/MongoOrderStatusHistoryServiceImpl.java [hunk 12]
```diff
@@ -24,7 +25,7 @@ public List<MongoOrderStatusHistory> findUnProcessedOrderStatusHistories() {
   }
 
   @Override
-  public void updateBatchToProcessed(List<ObjectId> ids) {
+  public void updateBatchToProcessed(HashSet<ObjectId> ids) {
```
Comment (reynoldmorel) [inline src/main/java/com/ea/restaurant/service/impl/MongoOrderStatusHistoryServiceImpl.java:28]: Use the Set interface instead

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/main/java/com/ea/restaurant/repository/MongoOrderStatusHistoryRepository.java (hunk 10 above)
Comment (eapg) [reply src/main/java/com/ea/restaurant/repository/MongoOrderStatusHistoryRepository.java:16]: Done.

File: src/main/java/com/ea/restaurant/service/MongoOrderStatusHistoryService.java (hunk 11 above)
Comment (eapg) [reply src/main/java/com/ea/restaurant/service/MongoOrderStatusHistoryService.java:11]: Done.

File: src/main/java/com/ea/restaurant/service/impl/MongoOrderStatusHistoryServiceImpl.java (hunk 12 above)
Comment (eapg) [reply src/main/java/com/ea/restaurant/service/impl/MongoOrderStatusHistoryServiceImpl.java:28]: Done.

**Review by reynoldmorel — APPROVED** [review]

**Pull request merged**